shell/platform/tizen/channels/text_input_channel.cc

```diff
@@ -40,252 +40,93 @@ constexpr char kTextKey[] = "text";
 
 TextInputChannel::TextInputChannel(BinaryMessenger* messenger,
                                    FlutterTizenEngine* engine)
     : channel_(std::make_unique<MethodChannel<rapidjson::Document>>(
           messenger,
           kChannelName,
           &JsonMethodCodec::GetInstance())),
-      engine_(engine) {
+      input_method_context_(std::make_unique<TizenInputMethodContext>(engine)) {
   channel_->SetMethodCallHandler(
       [this](const MethodCall<rapidjson::Document>& call,
              std::unique_ptr<MethodResult<rapidjson::Document>> result) {
         HandleMethodCall(call, std::move(result));
       });
 
-  ecore_imf_init();
-  // Register IMF callbacks
-  if (ecore_imf_context_default_id_get()) {
-    imf_context_ = ecore_imf_context_add(ecore_imf_context_default_id_get());
-  } else if (GetImfMethod() != nullptr) {
-    imf_context_ = ecore_imf_context_add(GetImfMethod());
-  }
-  if (imf_context_) {
-    ecore_imf_context_client_window_set(
-        imf_context_,
-        reinterpret_cast<void*>(engine_->renderer->GetWindowId()));
-    RegisterIMFCallback();
-  } else {
-    FT_LOGE("Failed to create imfContext");
-  }
+  // Set input method callbacks
+  input_method_context_->SetOnCommitCallback([this](std::string str) -> void {
+    FT_LOGI("OnCommit str[%s]", str.data());
+    text_editing_context_.edit_status_ = EditStatus::kCommit;
+    ConsumeLastPreedit();
+    active_model_->AddText(str);
+    SendStateUpdate(*active_model_);
+  });
+
+  input_method_context_->SetOnPreeditCallback(
+      [this](std::string str, int cursor_pos) -> void {
+        text_editing_context_.edit_status_ = EditStatus::kPreeditStart;
+        if (str.compare("") == 0) {
+          text_editing_context_.edit_status_ = EditStatus::kPreeditEnd;
+        }
+
+        if (text_editing_context_.edit_status_ == EditStatus::kPreeditStart ||
+            (text_editing_context_.edit_status_ == EditStatus::kPreeditEnd &&
+             // For tv, fix me
+             text_editing_context_.last_handled_ecore_event_keyname_.compare(
+                 "Return") != 0)) {
+          text_editing_context_.last_handled_ecore_event_keyname_ = "";
+          ConsumeLastPreedit();
+        }
+
+        text_editing_context_.has_preedit_ = false;
+        if (text_editing_context_.edit_status_ == EditStatus::kPreeditStart) {
+          text_editing_context_.preedit_start_pos_ =
+              active_model_->selection().base();
+          active_model_->AddText(str);
+          text_editing_context_.preedit_end_pos_ =
+              active_model_->selection().base();
+          text_editing_context_.has_preedit_ = true;
+          SendStateUpdate(*active_model_);
+          FT_LOGI("preedit start pos[%d], preedit end pos[%d]",
+                  text_editing_context_.preedit_start_pos_,
+                  text_editing_context_.preedit_end_pos_);
+        }
+      });
+
+  input_method_context_->SetOnInputPannelStateChangedCallback(
+      [this](int state) {
```

Comment:
Not specifying the exact type (`Ecore_IMF_Input_Panel_State`)?

Reply:
I followed the definition below.

``` c++
EAPI void ecore_imf_context_input_panel_event_callback_add(Ecore_IMF_Context *ctx, Ecore_IMF_Input_Panel_Event type, void (*func) (void *data, Ecore_IMF_Context *ctx, int value), const void *data);

```